### ISSUE TITLE
[vm] super_bounded_types_test Crash on simarm

### ISSUE THREAD

**Issue by dcharkes** [dart-lang/sdk#46572]
There are new test failures on [Prepare to publish meta 2.0.0...[vm] Prefix HOST_OS_* and TARGET_OS_* with DART_](https://github.com/dart-lang/sdk/compare/aa9201b76b21ff1d01494f692dcbc51492320b20~...0f2e58367d3f054fa3aeeb42bff68d7e8197dbcb).

The tests
```
language/generic/super_bounded_types_test Crash (expected Pass)
language/generic/super_bounded_types_test Crash (expected Pass)
```
are failing on configurations
```
dartk-strong-linux-release-simarm
dartk-weak-asserts-linux-release-simarm
```

```
--- Command "vm_compile_to_kernel" (took 01.000394s):
DART_CONFIGURATION=ReleaseSIMARM /b/s/w/ir/pkg/vm/tool/gen_kernel --no-aot --platform=out/ReleaseSIMARM/vm_platform_strong.dill -o /b/s/w/ir/out/ReleaseSIMARM/generated_compilations/dartk-strong-linux-release-simarm/tests_language_generic_super_bounded_types_test/out.dill /b/s/w/ir/tests/language/generic/super_bounded_types_test.dart --packages=/b/s/w/ir/.packages -Ddart.vm.product=false --sound-null-safety

exit code:
0

--- Command "vm" (took 686ms):
DART_CONFIGURATION=ReleaseSIMARM out/ReleaseSIMARM/dart --sound-null-safety --ignore-unrecognized-flags --packages=/b/s/w/ir/.packages /b/s/w/ir/out/ReleaseSIMARM/generated_compilations/dartk-strong-linux-release-simarm/tests_language_generic_super_bounded_types_test/out.dill

exit code:
-6

stderr:
===== CRASH =====
si_signo=Segmentation fault(11), si_code=1, si_addr=0xfffffff8
version=2.14.0-edge.0f2e58367d3f054fa3aeeb42bff68d7e8197dbcb (be) (Sat Jul 3 01:22:45 2021 +0000) on "linux_simarm"
pid=5040, thread=5055, isolate_group=main(0x345c800), isolate=main(0x3476880)
isolate_instructions=1eca4c0, vm_instructions=1eca4c0
Stack dump aborted because GetAndValidateThreadStackBounds failed.

--- Re-run this test:
python3 tools/test.py -n dartk-strong-linux-release-simarm language/generic/super_bounded_types_test
```

https://dart-ci.appspot.com/log/vm-kernel-nnbd-linux-release-simarm/dartk-strong-linux-release-simarm/259/language/generic/super_bounded_types_test

https://dart-ci.appspot.com/log/vm-kernel-nnbd-linux-release-simarm/dartk-weak-asserts-linux-release-simarm/259/language/generic/super_bounded_types_test

https://ci.chromium.org/ui/p/dart/builders/ci.sandbox/vm-kernel-nnbd-linux-release-simarm/259/overview

None of the 34 commits are obvious candidates. We'll have to bisect.

**Comment by dcharkes** [comment]
Succeeding on aa9201b76b21ff1d01494f692dcbc51492320b20.
Failing on d25fdc43ba4aa9d94f0909699a846b857e92186a.

Cause: https://dart-review.googlesource.com/c/sdk/+/203766

**Comment by sstrickl** [comment]
The majority of that CL deals with precompiled files, and this is not a precompiled test. Thus, it must be the changes in utils to `Utils::Is{,U,AbsoluteU}Int<T>(N, value)` to allow `N >= sizeof(T)`. Let's see...

**Comment by sstrickl** [comment]
So tried changing the definition of `IsAbsoluteUint(N, value)` back to the old one (where it uses `IsUint(N, -value)` if negative, else `IsUint(N, value)`)  and that works. So did some printing of when that definition mismatches with the new one (`IsInt(N + 1, value)`) and got:

```
IsAbsoluteUint<int32_t>(12, 0xfffff000) != IsInt<int32_t>(13, 0xfffff000)
```

That is, `IsUint(12, 0x0001000)` is false, which is correct, but the old definition doesn't fit the latter part of the description of `IsAbsoluteUint`:

>   // Check whether the magnitude of value fits in N bits, i.e., whether an
  // (N+1)-bit sign-magnitude representation can hold value.

That is, the old definition of `IsAbsoluteUint` fails for the minimum value `v_min`, where the bit length of `v_min` is N but the bit length of `-v_min` is N + 1. However, the latter addition  to the description would suggest it succeeds since `v_min` can be stored in an signed int of size N + 1.

In particular, the mismatch is happening at this line: https://github.com/dart-lang/sdk/blob/d25fdc43ba4aa9d94f0909699a846b857e92186a/runtime/vm/compiler/assembler/assembler_arm.cc#L2379

called from `StoreToOffset`: https://github.com/dart-lang/sdk/blob/d25fdc43ba4aa9d94f0909699a846b857e92186a/runtime/vm/compiler/assembler/assembler_arm.cc#L2879

Namely, the call being: `StoreToOffset(R0, R11, -4096, compiler::kFourBytes, AL)`.

  So what that means is normally this store is being converted to an `AddImmediate` + following store  (because `Address::CanHoldStoreOffset` fails), but with this change, the immediate is being encoded directly via conversion to an `Address`. The address encoding generated is `0x10b1000`, which then gets passed to:

```
EmitMemOp(AL, false,  false, R0, <address>)
```

For a final instruction coding of `0xe50b1000`. That is, the 13-bit encoding of the magnitude (4096) is bleeding into the Rt field and causing it to store to R1, not R0.

That's because the offset isn't prepended with the `U` bit (bit 23) to create a signed int, but the `U` bit determines whether or not the offset is added or subtracted (thus with an offset of 0, either `U` bit value gives the same result).

Long story short, the description for the function needs adjusting as well as  restoring the old behavior, and am  doing that now on a new CL.

**Comment by eernstg** [comment]
@sstrickl wrote:
> Long story

Long, but cool! :smile:

**Comment by dcharkes** [comment]
Nice one @sstrickl!

**Comment by sstrickl** [comment]
Note that after further reflection, the original documentation was completely correct. It was my reading of "sign-magnitude" as "two's complement" (which stores sign and magnitude information, but is not the same) that was wrong :)
